examples/aggregator/module/rust/src/lib.rs

```diff
@@ -22,76 +22,117 @@
 //! Clients invoke the module by providing a vector of non-negative numbers and get back an
 //! aggregated vector if an Oak Node has collected more samples than the predefined threshold.
 
-mod aggregator;
+mod data;
 mod proto;
 #[cfg(test)]
 mod tests;
 
-use aggregator::{Monoid, ThresholdAggregator};
-use log::info;
+use aggregator_common::ThresholdAggregator;
+use data::SparseVector;
+use log::{debug, error};
 use oak::grpc;
-use proto::aggregator::Vector;
-use proto::aggregator_grpc::{Aggregator, Dispatcher};
+use proto::aggregator::{Sample, SerializedSparseVector};
+use proto::aggregator_grpc::{Aggregator, AggregatorClient, Dispatcher};
 use protobuf::well_known_types::Empty;
+use std::collections::HashMap;
+use std::convert::{From, TryFrom};
 
-const SAMPLE_THRESHOLD: u64 = 3;
+/// Currently threshold value is hardcoded.
+const SAMPLE_THRESHOLD: u64 = 2;
 
-oak::entrypoint!(oak_main => {
-    oak::logger::init_default();
-    let node = AggregatorNode {
-        aggregator: ThresholdAggregator::<Vector>::new(SAMPLE_THRESHOLD),
-    };
-    Dispatcher::new(node)
-});
+type ThresholdAggregatorMap = HashMap<String, Option<ThresholdAggregator<SparseVector>>>;
 
-impl Monoid for Vector {
-    fn identity() -> Self {
-        Vector::new()
-    }
+/// Oak Node that collects aggregated data.
+pub struct AggregatorNode {
+    aggregators: ThresholdAggregatorMap,
+}
 
-    /// Adds two non-negative integer vectors elementwise. If vectors have different lengths, then
-    /// pads the smallest vector with zeros.
-    fn combine(&self, other: &Self) -> Self {
-        use itertools::EitherOrBoth::*;
-        use itertools::Itertools;
+impl AggregatorNode {
+    fn new() -> Self {
+        AggregatorNode {
+            aggregators: ThresholdAggregatorMap::new(),
+        }
+    }
 
-        Vector {
-            items: self
-                .items
-                .iter()
-                .zip_longest(other.items.iter())
-                .map(|p| match p {
-                    Both(l, r) => *l + *r,
-                    Left(l) => *l,
-                    Right(r) => *r,
-                })
-                .collect(),
-            ..Default::default()
+    fn submit_sparse_vector(&mut self, bucket: String, svec: &SparseVector) -> Result<(), String> {
```

Comment:
This method name seems misleading; `submit` makes sense as the gRPC name from the client point of view, but having an internal method called `submit` that does not actually submit anything does not seem very accurate.

But I think this actually comes from this name:

https://github.com/project-oak/oak/blob/0bb76ad2b58e1a26c551e3805cf6089c846308e0/examples/aggregator/module/rust/src/aggregator.rs#L48

Perhaps it should be renamed to something else, but I don't have a good suggestion. A few random things that come to mind but do not seem ideal: `add`, `contribute`, `merge`.

@project-oak/core any suggestions?